File: services/geoapi/geoapi/HTTP/Handlers/GetTerrainObjectHandler.cs
﻿using System.Net;
using System.Threading.Tasks;
using geoapi.Storages;
using geoapi.Utils;

namespace geoapi.HTTP.Handlers
{
    public class GetTerrainObjectHandler : BaseHandler
    {

        private readonly IStorage storage;
        private readonly ITerrainObjectStore terrainObjectStore;

        public GetTerrainObjectHandler(IStorage storage, ITerrainObjectStore terrainObjectStore) : base("GET", "object")
        {
            this.storage = storage;
            this.terrainObjectStore = terrainObjectStore;
        }

        protected override async Task HandleRequestAsync(HttpListenerContext context)
        {
            var objectKey = context.Request.QueryString[ObjectKeyParameter];

            var agentKey = context.Request.QueryString[AgentKeyParameter];

            var agent = storage.GetAgent(agentKey);

            if (agent == null)
            {
                await context.Response.Send(404, "Object not found");
                return;
            }

            var tObject = terrainObjectStore.GetTerrainObject(objectKey);

            if (tObject == null)
            {
                await context.Response.Send(404, "Object not found");
                return;
            }

            await context.Response.Send(200, tObject.ToJson());
        }
    }
}

File: services/geoapi/geoapi/HTTP/Handlers/GetTerrainObjectHandler.vb
Imports System.Net
Imports System.Threading.Tasks
Imports geoapi.Storages
Imports geoapi.Utils

Namespace HTTP.Handlers
    Public Class GetTerrainObjectHandler
        Inherits BaseHandler

        Private ReadOnly agentStorage As IAgentStorage
        Private ReadOnly terrainObjectStore As ITerrainObjectStore

        Public Sub New(ByVal agentStorage As IAgentStorage, ByVal terrainObjectStore As ITerrainObjectStore)
            MyBase.New("GET", "object")
            Me.agentStorage = agentStorage
            Me.terrainObjectStore = terrainObjectStore
        End Sub

        Protected Overrides Async Function HandleRequestAsync(ByVal context As HttpListenerContext) As Task
            Dim objectKey = context.Request.QueryString(ObjectKeyParameter)
            Dim agentKey = context.Request.QueryString(AgentKeyParameter)
            Dim agent = agentStorage.GetAgent(agentKey)

            If agent Is Nothing Then
                Await context.Response.Send(404, "Agent not found")
                Return
            End If

            Dim tObject = terrainObjectStore.GetTerrainObject(objectKey)

            If tObject Is Nothing Then
                Await context.Response.Send(404, "Object not found")
                Return
            End If

            Await context.Response.Send(200, tObject.ToJson())
        End Function
    End Class
End Namespace
